ipa-core/src/protocol/ipa_prf/mod.rs

```diff
@@ -1,6 +1,232 @@
+use std::iter::{repeat, zip};
+
+use ipa_macros::Step;
+
+use crate::{
+    error::Error,
+    ff::{boolean::Boolean, boolean_array::BA64, CustomArray, Field, PrimeField, Serializable},
+    helpers::query::IpaQueryConfig,
+    protocol::{
+        context::{UpgradableContext, UpgradedContext},
+        ipa_prf::{
+            boolean_ops::convert_to_fp25519,
+            prf_eval::{eval_dy_prf, gen_prf_key},
+            prf_sharding::{
+                attribute_cap_aggregate, compute_histogram_of_users_with_row_count,
+                PrfShardedIpaInputRow,
+            },
+        },
+        RecordId,
+    },
+    report::OprfReport,
+    secret_sharing::{
+        replicated::{malicious::ExtendableField, semi_honest::AdditiveShare as Replicated},
+        WeakSharedValue,
+    },
+};
+
 mod boolean_ops;
-#[cfg(feature = "descriptive-gate")]
 pub mod prf_eval;
 pub mod prf_sharding;
 #[cfg(feature = "descriptive-gate")]
 pub mod shuffle;
+
+#[derive(Step)]
+pub(crate) enum Step {
+    ConvertFp25519,
+    EvalPrf,
+    ConvertInputRowsToPrf,
+}
+
+/// IPA OPRF Protocol
+///
+/// We return `Replicated<F>` as output.
+/// This protocol does following steps
+/// 1. Shuffles the input (TBD)
+/// 2. Converts boolean arrays of match keys to elliptical values
```

Comment:
That makes more sense actually. I think when we were discussing earlier, we didnt consider adding dummies hence thought it could be done in beginning or here. Updated